File: python/src/threads/agents/skills.py
"""Skills (spec/api.json `Skill`): host config, pinned by hash at
thread start. Only what the host passes to `agent(skills=...)` is a skill; a skill-shaped file in
the sandbox or repo is data and never loads as one (F4.3). Line 0 lists each name and
description; load_skill appends a body as a trusted `injected{source: skill}` (F4.1)."""

import re
from collections.abc import Sequence
from dataclasses import dataclass
from typing import Final

from pydantic import JsonValue

from threads._generated.tools_v1 import LoadSkillInput
from threads.agents.config import ConfigError
from threads.log import JsonObject, ToolSpec
from threads.log.digest import sha256_hex
from threads.loop.model import LookupResult, LookupUnknown
from threads.loop.tools import Dispatched, Invocation, Output, Reference, Termination
from threads.result import Err
from threads.tools.runner import parse

_NAME: Final = re.compile(r"[a-z][a-z0-9_]{0,63}")
"""events.v1 `$defs/Name`."""


@dataclass(frozen=True, slots=True)
class Skill:
    """spec/api.json `Skill`: listed by name and description; the body loads on demand."""

    name: str
    description: str
    body: str

    @property
    def version(self) -> str:
        """The SHA-256 of the body: what config_hash pins and the injection's origin names."""
        return sha256_hex(self.body.encode("utf-8"))


def checked(skills: Sequence[Skill]) -> tuple[Skill, ...]:
    """Raises ConfigError for a name that isn't a wire name or repeats, or a description that
    isn't one non-empty line."""
    for s in skills:
        if _NAME.fullmatch(s.name) is None:
            raise ConfigError("invalid_config", f"skill name {s.name!r} isn't a wire name")
        if not s.description.strip() or "\n" in s.description:
            raise ConfigError("invalid_config", f"skill {s.name}: one non-empty description line")
    names = [s.name for s in skills]
    if len(set(names)) != len(names):
        raise ConfigError("duplicate_name", f"skill names repeat: {names}")
    return tuple(skills)


def listing(skills: Sequence[Skill]) -> str:
    """The line-0 listing: names and descriptions only, never a body."""
    if not skills:
        return ""
    lines = (f"- {s.name}: {s.description}" for s in skills)
    return "Skills you can load with load_skill:\n" + "\n".join(lines)


def pinned(skills: Sequence[Skill]) -> list[JsonValue]:
    """What config_hash pins for each skill: a changed body is a changed config."""
    return [{"name": s.name, "description": s.description, "sha256": s.version} for s in skills]


class SkillLoader:
    """load_skill: host-side and read_only over the skills pinned at thread start."""

    def __init__(self, skills: Sequence[Skill]) -> None:
        self._skills = {s.name: s for s in skills}

    def invalid(self, spec: ToolSpec, input: JsonObject) -> str | None:
        parsed = parse(spec.name, input)
        return parsed.error if isinstance(parsed, Err) else None

    async def dispatch(self, call: Invocation) -> Dispatched:
        parsed = parse(call.spec.name, call.input)
        if isinstance(parsed, Err) or not isinstance(parsed.value, LoadSkillInput):
            raise AssertionError("a dispatched call was parsed first")
        name = parsed.value.name
        skill = self._skills.get(name)
        if skill is None:
            return Output(f"not_found: no skill named {name}; only the listed skills exist", True)
        ref = Reference("skill", skill.name, skill.version, skill.body)
        return Output(f"loaded skill {name}", False, None, (ref,))

    async def lookup(self, call: Invocation) -> LookupResult[str]:
        return LookupUnknown("load_skill has no lookup")

    async def terminate(self, call: Invocation) -> Termination:
        return "unknown"

    def provider_now(self) -> int | None:
        return None

File: typescript/packages/core/src/agent/skills.ts
import { sha256Hex } from "../hash";
import { Name, ToolSpec } from "../log";
import type { ToolImpl } from "../loop/types";
import { entry } from "../tools/catalog";
import { LoadSkillInput } from "../tools/memory-inputs";
import { ConfigError } from "./errors";
import { jsonSchema } from "./tool";

// Skills: host config, pinned by config_hash at thread start.
// Line 0 lists each name and description; load_skill appends a body from this pinned set only,
// never from the sandbox or repo (F4.3). Compaction restores loaded skills (loop/restore.ts).

/** spec/api.json Skill. */
export type Skill = {
  readonly name: string;
  readonly description: string;
  readonly body: string;
};

/** Config is a boundary: names are unique Names, descriptions one non-empty line. */
export function checkSkills(skills: readonly Skill[]): void {
  const names = skills.map((s) => s.name);
  const bad = skills.find(
    (s) =>
      !Name.safeParse(s.name).success ||
      s.description.trim() === "" ||
      /[\r\n]/.test(s.description),
  );
  if (bad !== undefined)
    throw new ConfigError(
      "invalid_config",
      `skill ${JSON.stringify(bad.name)} needs a lowercase name and a one-line description`,
    );
  const twice = names.find((n, i) => names.indexOf(n) !== i);
  if (twice !== undefined)
    throw new ConfigError("duplicate_name", `two skills are named ${twice}`);
}

/** The listing paragraph of line 0's system text; none without skills. */
export function skillListing(skills: readonly Skill[]): readonly string[] {
  if (skills.length === 0) return [];
  const lines = skills.map((s) => `- ${s.name}: ${s.description}`);
  return [["Skills you can load with load_skill:", ...lines].join("\n")];
}

/** What config_hash pins: the listing plus each body's hash (bodies aren't in line 0). */
export function skillPins(
  skills: readonly Skill[],
): readonly Record<string, string>[] {
  return skills.map((s) => ({
    name: s.name,
    description: s.description,
    sha256: sha256Hex(s.body),
  }));
}

export function skillSpecs(skills: readonly Skill[]): readonly ToolSpec[] {
  if (skills.length === 0) return [];
  const e = entry("load_skill");
  return [
    ToolSpec.parse({
      name: e.name,
      description: e.description,
      input_schema: jsonSchema(e.name, e.input),
      effect_class: "read_only",
    }),
  ];
}

/** load_skill over the pinned set: the body arrives as injected trusted instruction. */
export function skillTools(skills: readonly Skill[]): readonly ToolImpl[] {
  return skillSpecs(skills).map((spec) => ({
    spec,
    input: LoadSkillInput,
    run: async (raw) => {
      const { name } = LoadSkillInput.parse(raw);
      const skill = skills.find((s) => s.name === name);
      if (skill === undefined)
        return {
          kind: "done",
          output: `not_found: no skill named ${name}; only the listed skills exist`,
          isError: true,
        };
      return {
        kind: "done",
        output: `loaded skill ${name}`,
        isError: false,
        inject: [
          {
            source: "skill",
            trust: "trusted_instruction",
            origin: { id: name, version: sha256Hex(skill.body) },
            text: skill.body,
          },
        ],
      };
    },
  }));
}
